src/Stratis.Bitcoin/P2P/PeerAddressManager.cs

```diff
@@ -136,22 +150,51 @@ public void SavePeers()
         /// <inheritdoc/>
         public void AddPeer(NetworkAddress networkAddress, IPAddress source, PeerIntroductionType peerIntroductionType)
         {
+            Guard.Assert(peerIntroductionType != PeerIntroductionType.Connect);
+
             if (networkAddress.Endpoint.Address.IsRoutable(true) == false)
                 return;
 
             var peerToAdd = PeerAddress.Create(networkAddress, source, peerIntroductionType);
             this.Peers.TryAdd(peerToAdd.NetworkAddress.Endpoint, peerToAdd);
         }
 
+        /// <inheritdoc/>
+        public void AddPeerWithConnectArg(NetworkAddress networkAddress, IPAddress source)
```

Comment:
I think instead of adding a new method you should use the overload `this.Peers.AddOrUpdate` in the method `AddPeer` and update `peer.PeerIntroductionType` in case of an update.